Applications/SpreadSheet/SpreadSheet.cpp

```diff
@@ -0,0 +1,294 @@
+        m_visited_cells_in_update.set(it.value);
+        if (it.value->dirty) {
+            // Re-evaluate the cell value, if any.
+            it.value->update({});
+        }
+    }
+
+    m_visited_cells_in_update.clear();
+}
+
+void Sheet::update(Cell& cell)
+{
+    if (has_been_visited(&cell))
+        return;
+
+    m_visited_cells_in_update.set(&cell);
+    cell.update({});
+}
+
+JS::Value Sheet::evaluate(const StringView& source, Cell* on_behalf_of)
+{
+    TemporaryChange cell_change { m_current_cell_being_evaluated, on_behalf_of };
+
+    auto parser = JS::Parser(JS::Lexer(source));
+    if (parser.has_errors())
+        return JS::js_undefined();
+
+    auto program = parser.parse_program();
+    m_interpreter->run(m_interpreter->global_object(), program);
+    if (m_interpreter->exception()) {
+        auto exc = m_interpreter->exception()->value();
+        m_interpreter->clear_exception();
+        return exc;
+    }
+
+    auto value = m_interpreter->last_value();
+    if (value.is_empty())
+        return JS::js_undefined();
+    return value;
+}
+
+void Cell::update_data()
+{
+    TemporaryChange cell_change { sheet->current_evaluated_cell(), this };
+    if (!dirty)
+        return;
+
+    dirty = false;
+    if (kind == Formula) {
+        if (!evaluated_externally)
+            evaluated_data = sheet->evaluate(data, this);
+    }
+
+    for (auto& ref : referencees) {
+        if (ref) {
+            ref->dirty = true;
+            ref->update();
+        }
+    }
+}
+
+void Cell::update()
+{
+    sheet->update(*this);
+}
+
+JS::Value Cell::js_data()
+{
+    if (dirty)
+        update();
+
+    if (kind == Formula)
+        return evaluated_data;
+
+    return JS::js_string(sheet->interpreter(), data);
+}
+
+Cell* Sheet::at(const StringView& name)
+{
+    auto pos = parse_cell_name(name);
+    if (pos.has_value())
+        return at(pos.value());
+
+    return nullptr;
+}
+
+Cell* Sheet::at(const Position& position)
+{
+    auto it = m_cells.find(position);
+
+    if (it == m_cells.end())
+        return nullptr;
+
+    return it->value;
+}
+
+Optional<Position> Sheet::parse_cell_name(const StringView& name)
+{
+    GenericLexer lexer(name);
+    auto col = lexer.consume_while([](auto c) { return is_alpha(c); });
+    auto row = lexer.consume_while([](auto c) { return is_alphanum(c) && !is_alpha(c); });
+
+    if (!lexer.is_eof() || row.is_empty() || col.is_empty())
+        return {};
+
+    return Position { col, row.to_uint().value() };
+}
+
+String Cell::source() const
+{
+    StringBuilder builder;
+    if (kind == Formula)
+        builder.append('=');
+    builder.append(data);
+    return builder.to_string();
+}
+
+// FIXME: Find a better way to figure out dependencies
+void Cell::reference_from(Cell* other)
+{
+    if (!other || other == this)
+        return;
+
+    if (!referencees.find([other](auto& ptr) { return ptr.ptr() == other; }).is_end())
+        return;
+
+    referencees.append(other->make_weak_ptr());
```

Comment:
`referencee` would be someone being referred *to*.
In this case, they are `referencers`, but perhaps `referencing_cells` would be a bit easier to brain-parse? :)